components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodPortForwardingService.kt

```diff
@@ -0,0 +1,136 @@
+
+        val portsStatusRequest = Status.PortsStatusRequest.newBuilder().setObserve(true).build()
+
+        val portsStatusResponseObserver = object :
+                ClientResponseObserver<Status.PortsStatusRequest, Status.PortsStatusResponse> {
+            override fun beforeStart(request: ClientCallStreamObserver<Status.PortsStatusRequest>) {
+                project.lifetime.onTerminationOrNow { request.cancel("gitpod: Project terminated.", null) }
+            }
+            override fun onNext(response: Status.PortsStatusResponse) { handlePortStatusResponse(response) }
+            override fun onCompleted() { completableFuture.complete(null) }
+            override fun onError(throwable: Throwable) { completableFuture.completeExceptionally(throwable) }
+        }
+
+        statusServiceStub.portsStatus(portsStatusRequest, portsStatusResponseObserver)
+
+        return completableFuture
+    }
+
+    private fun handlePortStatusResponse(response: Status.PortsStatusResponse) {
+        val portNumberToServedStatusMap = mutableMapOf<Int, Boolean>()
+
+        for (port in response.portsList) {
+            val isServed = (port.exposed != null || port.tunneled != null) && port.served
+
+            portNumberToServedStatusMap[port.localPort] = isServed
+        }
+
+        updateForwardedPortsList(portNumberToServedStatusMap)
+    }
+
+    private fun updateForwardedPortsList(portNumberToServedStatusMap: Map<Int, Boolean>) {
+        val portForwardingManager = PortForwardingManager.getInstance(project)
+        val forwardedPortsList = portForwardingManager.getForwardedPortsWithLabel(FORWARDED_PORT_LABEL)
+
+        portNumberToServedStatusMap.forEach { (hostPort, isServed) ->
+            if (isServed && !forwardedPortsList.containsKey(hostPort)) {
+                val portEventsProcessor = object : PortEventsProcessor {
+                    override fun onPortForwarded(hostPort: Int, clientPort: Int) {
+                        portsService.setForwardedPort(hostPort, clientPort)
+                        thisLogger().info("gitpod: Forwarded port $hostPort to client's port $clientPort.")
+                    }
+
+                    override fun onPortForwardingEnded(hostPort: Int) {
+                        thisLogger().info("gitpod: Finished forwarding port $hostPort.")
+                    }
+
+                    override fun onPortForwardingFailed(hostPort: Int, reason: String) {
+                        thisLogger().error("gitpod: Failed to forward port $hostPort: $reason")
+                    }
+                }
+
+                val portInfo = ForwardedPortInfo(
+                        hostPort,
+                        RdPortType.HTTP,
+                        FORWARDED_PORT_LABEL,
+                        emptyList(),
+                        portEventsProcessor
+                )
+
+                portForwardingManager.forwardPort(portInfo)
```

Comment:
How about move forward behavior into `portsService.setForwardedPort`? Is it because of only latest can use it? Will we do it in the future (move code)?

Reply:
Exactly! We will move it to the "common" package when the Stable IDE version is compatible with and a few remaining issues get resolved! ✌️